src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/ModelBinderFactory.cs

```diff
@@ -86,29 +106,25 @@ private IModelBinder CreateBinderCore(DefaultModelBinderProviderContext provider
             // If we're currently recursively building a binder for this type, just return
             // a PlaceholderBinder. We'll fix it up later to point to the 'real' binder
             // when the stack unwinds.
-            var stack = providerContext.Stack;
-            for (var i = 0; i < stack.Count; i++)
+            var collection = providerContext.Collection;
+
+            IModelBinder binder;
+            if (collection.TryGetValue(key, out binder))
             {
-                var entry = stack[i];
-                if (key.Equals(entry.Key))
+                if (binder != null)
```

Comment:
Also can `providerContext.Collection` be removed? That is, do we need more than the `_cache`?